### PR TITLE
Add a StackDriver exporter for stats using a Pull model

### PR DESCRIPTION
**Pull request by songy23**
Per our discussion yesterday, we're not going to support a push model for exporting ViewData in Stats library in the first release. Instead, the StackDriver exporter will be more like an independent agent: users need to register views against exporter first, and then the exporter will poll stats from the Stats library for all registered views and upload it to StackDriver periodically.

### CODE REVIEW

File: exporters/stats_stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverExporter.java [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+
+  /**
+   * Register a {@link View} against this exporter, and upload it as a {@link MetricDescriptor} to
+   * StackDriver.
+   *
+   * @param view the {@code View} to be registered.
+   */
+  public void registerView(View view) {
+    workerThread.registerView(view);
+    // This is required only for custom view definitions. Canonical views should be pre-registered.
```
Comment (bogdandrutu) [inline exporters/stats_stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverExporter.java]: How do we enforce this?

File: exporters/stats_stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverExportUtils.java [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+    builder.setValueType(createValueType(view.getAggregation(), view.getMeasure()));
+    return builder.build();
+  }
+
+  // Construct a name for a MetricDescriptor
+  // Name format refers to cloud.google.com/monitoring/api/ref_v3/rest/v3/projects.metricDescriptors
+  private static String createMetricDescriptorName(
+      String projectId, String viewName) {
+    return String.format(
+        "projects/%s/metricDescriptors/custom.googleapis.com%%2Fcensus%%2F%s%%2F%s",
```
Comment (bogdandrutu) [inline exporters/stats_stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverExportUtils.java]: See here for name:
https://cloud.google.com/monitoring/api/ref_v3/rest/v3/projects.metricDescriptors/create

return String.format(projects/%s, project_id) -> also you can cache this because it is constant

File: exporters/stats_stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverExportUtils.java [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+        projectId, NAMESPACE, viewName);
+  }
+
+  // Construct a MetricDescriptor.Type. From the documentation:
+  // cloud.google.com/monitoring/api/ref_v3/rest/v3/projects.metricDescriptors
+  // For now only support custom metrics.
+  private static String createMetricType(String viewName) {
+    return String.format("custom.googleapis.com/census/%s/%s", NAMESPACE, viewName);
```
Comment (bogdandrutu) [inline exporters/stats_stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverExportUtils.java]: "custom.googleapis.com/opencensus/%s", viewName?

File: exporters/stats_stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverExportUtils.java [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+    }
+
+    MetricDescriptor.Builder builder = MetricDescriptor.newBuilder();
+    String name = view.getName().asString();
+    builder.setName(createMetricDescriptorName(projectId, name));
+    builder.setType(createMetricType(name));
+    builder.setDescription(view.getDescription());
+    builder.setUnit(view.getMeasure().getUnit());
+    // TODO(songya): do we want a different display name?
+    builder.setDisplayName(NAMESPACE + "/" + name);
```
Comment (bogdandrutu) [inline exporters/stats_stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverExportUtils.java]: Better to use:
"Opencensus " + name

Comment (songy23) [reply exporters/stats_stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverExportUtils.java]: Sounds good, updated.

File: exporters/stats_stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverExportUtils.java (hunk 2 above)
Comment (songy23) [reply exporters/stats_stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverExportUtils.java]: Done.

File: exporters/stats_stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverExportUtils.java (hunk 3 above)
Comment (songy23) [reply exporters/stats_stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverExportUtils.java]: Updated

File: exporters/stats_stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverExporter.java (hunk 1 above)
Comment (songy23) [reply exporters/stats_stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverExporter.java]: We can check if the view is one of [RpcViewConstants](https://github.com/census-instrumentation/opencensus-java/blob/master/core/src/main/java/io/opencensus/stats/RpcViewConstants.java), if so then it doesn't need to be posted to StackDriver. This is not needed for now, since we haven't defined canonical metrics on StackDriver. I've added a TODO here.

**Comment by adriancole** [comment]
Curious.. are you thinking the same for trace data?

**Comment by songy23** [comment]
@adriancole probably not, since views need to be registered as metrics first in order to export stats to Stackdriver, while span data do not. In addition, we're considering modifying Stats APIs to support a push model for stats and having an export component like trace. That's probably for next release though.

File: exporters/stats_stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverExportUtils.java [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+    builder.setValueType(valueType);
+    return builder.build();
+  }
+
+  // Construct a MetricKind from an AggregationWindow
+  @VisibleForTesting
+  static MetricKind createMetricKind(AggregationWindow window) {
+    return window.match(
+        Functions.returnConstant(MetricKind.CUMULATIVE),               // Cumulative
+        // We don't support exporting Interval stats to StackDriver in v0.1
```
Comment (bogdandrutu) [inline exporters/stats_stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverExportUtils.java]: I would rephrase it as:
// We don't support exporting Interval stats to StackDriver in this version.

Same everywhere. The current version of the library is 0.8.0 I know that this is the first stats release but it is confusing.

File: exporters/stats_stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverExporter.java [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+      }
+    }
+
+    // Polls ViewData from Stats library for all registered views, and upload them as TimeSeries to
+    // StackDriver.
+    private void export() {
+      CreateTimeSeriesRequest.Builder builder = CreateTimeSeriesRequest.newBuilder();
+      synchronized (monitor) {
+        for (View.Name view : registeredViews.keySet()) {
+          builder.addAllTimeSeries(createTimeSeriesList(viewManager.getView(view), projectId));
```
Comment (bogdandrutu) [inline exporters/stats_stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverExporter.java]: To avoid holding the lock for too long better to just extract the ViewData, add to a list then outside the lock create the proto.

File: exporters/stats_stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverExportUtils.java [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+    builder.setMetricKind(createMetricKind(view.getWindow()));
+    builder.setValueType(createValueType(view.getAggregation(), view.getMeasure()));
+    return builder.build();
+  }
+
+  // Construct a MetricDescriptor.Type. From the documentation:
+  // cloud.google.com/monitoring/api/ref_v3/rest/v3/projects.metricDescriptors
+  // For now only support custom metrics.
+  private static String createMetricType(String viewName) {
+    return String.format("custom.googleapis.com/opencensus/%s", viewName);
```
[the remark this answers is not in the document]
Comment (songy23) [reply exporters/stats_stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverExportUtils.java]: Done.

File: exporters/stats_stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverExportUtils.java (hunk 5 above)
Comment (songy23) [reply exporters/stats_stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverExportUtils.java]: Done.

File: exporters/stats_stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverExporter.java (hunk 6 above)
Comment (songy23) [reply exporters/stats_stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverExporter.java]: Updated

File: exporters/stats_stackdriver/src/test/java/io/opencensus/exporter/stats/stackdriver/StackdriverExporterTest.java [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+        CumulativeData.create(Timestamp.fromMillis(100), Timestamp.fromMillis(200)));
+    doNothing().when(mockViewManager).registerView(view);
+    doReturn(viewData).when(mockViewManager).getView(VIEW_NAME);
+
+    StackdriverExporter exporter = new StackdriverExporter(
+        PROJECT_ID, new FakeMetricServiceClient(mockStub), ONE_SECOND, mockViewManager);
+    exporter.registerView(view);
+
+    verify(mockStub, times(1)).createMetricDescriptorCallable();
+    verify(mockStub, timeout(1000).times(1)).createTimeSeriesCallable();
```
Comment (sebright) [inline exporters/stats_stackdriver/src/test/java/io/opencensus/exporter/stats/stackdriver/StackdriverExporterTest.java]: Does the timeout need to match the export interval for this test to pass?  It would help to have a comment explaining the timing in this test.

File: exporters/stats_stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverExporter.java [hunk 9]
```diff
@@ -0,0 +1,277 @@
...
+      MetricServiceClient metricServiceClient,
+      Duration exportInterval,
+      ViewManager viewManager) {
+    checkNotNull(exportInterval, "exportInterval");
+    checkArgument(exportInterval.compareTo(ZERO) > 0, "Duration must be positive");
+    this.projectId = checkNotNull(projectId, "projectId");
+    this.metricServiceClient = metricServiceClient;
+    this.workerThread = new WorkerThread(
+        projectId, metricServiceClient, exportInterval, viewManager);
+    workerThread.start();
```
Comment (sebright) [inline exporters/stats_stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverExporter.java]: It would be better to not start the thread in the constructor, because the "this" reference could escape from the constructor if WorkerThread were ever made non-static.  You could start it in `createInternal` instead.

File: exporters/stats_stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverExporter.java [hunk 10]
```diff
@@ -0,0 +1,277 @@
...
+   * Creates a StackdriverExporter for an explicit project ID and using explicit credentials.
+   *
+   * <p>Only one Stackdriver exporter can be created at any point.
+   *
+   * @param credentials a credentials used to authenticate API calls.
+   * @param projectId the cloud project id.
+   * @param exportInterval the interval to export stats.
+   * @throws IllegalStateException if a Stackdriver exporter already exists.
+   */
+  public static StackdriverExporter createWithCredentialsAndProjectId(
```
[the remark this answers is not in the document]
Comment (sebright) [reply exporters/stats_stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverExporter.java]: I thought about this a bit more, and I think I prefer making `registerView` static, if it is not possible to export to StackDriver with two different values for credentials, project ID, and time interval.  An instance method could cause users to think that they can register different sets of views on different instances of StackdriverExporter.  A static method makes it clear that registered views are global state.

File: exporters/stats_stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverStatsExporter.java [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+
+  /**
+   * Register a {@link View} against this exporter, and upload it as a {@link MetricDescriptor} to
+   * StackDriver.
+   *
+   * @param view the {@code View} to be registered.
+   * @throws IllegalStateException if a Stackdriver stats exporter has not been created yet.
+   */
+  public static synchronized void registerView(View view) {
+    synchronized (monitor) {
```
[the remark this answers is not in the document]
Comment (songy23) [reply exporters/stats_stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverStatsExporter.java]: Done.

**Comment by bogdandrutu** [comment]
Very small request. Probably better to move the exporter into:
exporters/stats/stackdriver instead of exporters/stats_stackdriver

See this PR https://github.com/census-instrumentation/opencensus-java/pull/719 for trace.

**Comment by songy23** [comment]
Sure, updated.

File: exporters/stats/stackdriver/README.md [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+For Gradle add to your dependencies: TODO
+
+#### Register the exporter
+
+This uses the default configuration for authentication and a given project ID.
+
+```java
+public class MyMainClass {
+  public static void main(String[] args) {
+    StackdriverStatsExporter.createWithProjectId("MyStackdriverProjectId", Duration.create(10, 0));
```
Comment (adriancole) [inline exporters/stats/stackdriver/README.md]: Add comment or otherwise what the duration is.. ex if interval do

// install the exporter with a 10 second window for aggregation (or whatever it is)

File: exporters/stats/stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverExportUtils.java [hunk 13]
```diff
@@ -0,0 +1,301 @@
...
+        // TODO(songya): We don't support exporting Interval stats to StackDriver in this version.
+        Functions.returnConstant(MetricKind.UNRECOGNIZED), // Interval
+        Functions.returnConstant(MetricKind.UNRECOGNIZED));
+  }
+
+  // Construct a MetricDescriptor.ValueType from an Aggregation and a Measure
+  @VisibleForTesting
+  static MetricDescriptor.ValueType createValueType(
+      Aggregation aggregation, final Measure measure) {
+    return aggregation.match(
```
Comment (adriancole) [inline exporters/stats/stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverExportUtils.java]: Maybe revisit this with a builder or some fluent api as this looks cryptic in practice

File: exporters/stats/stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverExportUtils.java [hunk 14]
```diff
@@ -0,0 +1,301 @@
...
+  static Point createPoint(
+      AggregationData aggregationData, AggregationWindowData windowData, Aggregation aggregation) {
+    Point.Builder builder = Point.newBuilder();
+    builder.setInterval(createTimeInterval(windowData));
+    builder.setValue(createTypedValue(aggregation, aggregationData));
+    return builder.build();
+  }
+
+  // Convert AggregationWindowData to TimeInterval, currently only support CumulativeData.
+  @VisibleForTesting
```
Comment (adriancole) [inline exporters/stats/stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverExportUtils.java]: Visible for test or only used in test? If not only used in tests avoid anonymous types (extract or constant and/or declare class with nice tostring for debug)

Comment (adriancole) [reply exporters/stats/stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverExportUtils.java]: If only used in test move to test tree

File: exporters/stats/stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverStatsExporter.java [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+   *
+   * @param exportInterval the interval between pushing stats to StackDriver.
+   * @throws IllegalStateException if a Stackdriver exporter is already created.
+   */
+  public static void create(Duration exportInterval) throws IOException {
+    checkNotNull(exportInterval, "exportInterval");
+    createInternal(null, ServiceOptions.getDefaultProjectId(), exportInterval);
+  }
+
+  private static void createInternal(
```
Comment (adriancole) [inline exporters/stats/stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverStatsExporter.java]: Comment create internal instead of just hidden ctor to enforce singleton

File: exporters/stats/stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverStatsExporter.java [hunk 16]
```diff
@@ -0,0 +1,205 @@
...
+      @Nullable Credentials credentials, String projectId, Duration exportInterval)
+      throws IOException {
+    synchronized (monitor) {
+      checkState(exporter == null, "Stackdriver stats exporter is already created.");
+      MetricServiceClient metricServiceClient;
```
Comment (adriancole) [inline exporters/stats/stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverStatsExporter.java]: Unless this has side effect or we expect tons of calls i would initialize this outside the lock. If there is a good reason to init this inside lock comment why

File: exporters/stats/stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverStatsExporter.java [hunk 17]
```diff
@@ -0,0 +1,205 @@
...
+                    .build());
+      }
+      exporter =
+          new StackdriverStatsExporter(
+              projectId, metricServiceClient, exportInterval, Stats.getViewManager());
+      exporter.workerThread.start();
+    }
+  }
+
+  // Method for setting exporter to a fake exporter or null (reset) for unit tests.
```
Comment (adriancole) [inline exporters/stats/stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverStatsExporter.java]: One crap thing about visible for tests is hard to tell only for tests or visible also for tests. I will just mention anyting that doesnt need to go in src/main should not be in src/main. Usually i put any heavy logic in test tree and only poke holes or allow minimum hooks in main. That may be the case here

File: exporters/stats/stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/WorkerThread.java [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+import io.opencensus.stats.View;
+import io.opencensus.stats.ViewData;
+import io.opencensus.stats.ViewManager;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.concurrent.GuardedBy;
+
+/** Worker {@code Thread} that polls ViewData from Stats library and batch export to StackDriver. */
+final class WorkerThread extends Thread {
```
Comment (adriancole) [inline exporters/stats/stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/WorkerThread.java]: If this is to be run as daemon mention? Otherwise say something like interrupt to stop loop (as there is no other way to cancel implemented here).

File: exporters/stats/stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverStatsExporter.java (hunk 17 above)
Comment (songy23) [reply exporters/stats/stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverStatsExporter.java]: Sorry for the confusion. This method is only for tests, however it cannot go into src/tests since it needs to access a private field, which is also guarded by a private lock of this class.

File: exporters/stats/stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/WorkerThread.java (hunk 18 above)
Comment (songy23) [reply exporters/stats/stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/WorkerThread.java]: It's a daemon thread. Added comment here.

**Comment by songy23** [comment]
Per our discussion today, since we'll probably update to use a push model for `ViewData`, it's better not to have `registerView()` at the exporter level for backwards compatibility. Instead we'll need an additional field in `View` that determine whether it should be local or exported to exporters.

For this version, since Stackdriver exporter is using a pull model, we'll need another API `getExportedViews` in our library. Exporter needs to call `getExportedViews` to get those `ViewData` that should be exported, instead of registering them against the exporter.

File: exporters/stats/stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverStatsExporter.java (hunk 17 above)
Comment (dinooliva) [reply exporters/stats/stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverStatsExporter.java]: Does it make sense to call it unsafeSetExporter() then?

File: exporters/stats/stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/WorkerThread.java [hunk 19]
```diff
@@ -0,0 +1,118 @@
...
+      String projectId,
+      MetricServiceClient metricServiceClient,
+      Duration exportInterval,
+      ViewManager viewManager) {
+    this.scheduleDelayMillis = toMillis(exportInterval);
+    this.projectId = projectId;
+    this.metricServiceClient = metricServiceClient;
+    this.viewManager = viewManager;
+    setDaemon(true);
+    setName("ExportWorkerThread");
```
Comment (dinooliva) [inline exporters/stats/stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/WorkerThread.java]: Consider "StackdriverExporterWorkerThread" in case there are ever other exporters.

Comment (songy23) [reply exporters/stats/stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/WorkerThread.java]: Done.

File: exporters/stats/stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverStatsExporter.java (hunk 17 above)
Comment (songy23) [reply exporters/stats/stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverStatsExporter.java]: Sounds better, updated.

**Comment by songy23** [comment]
Thanks for review!
